File: src/cerebro_sdk/testing/analytics.py
"""Analytics testing harnesses for Cerebro SDK clients."""

from __future__ import annotations

from datetime import datetime
from typing import Iterable, Sequence

from cerebro_sdk.analytics import (
    IntegrationAccountSummary,
    IntegrationCoverageClient,
    IntegrationCoverageRecord,
    IntegrationScopeBreakdown,
    RuntimeEventAggregate,
    RuntimeHealthClient,
    RuntimeHealthRecord,
    RuntimeMetadataSnapshot,
)


def build_runtime_health_record(
    runtime: str,
    *,
    window_start: datetime,
    window_end: datetime,
    events: dict[str, RuntimeEventAggregate] | None = None,
    warnings: dict[str, RuntimeEventAggregate] | None = None,
    metadata: RuntimeMetadataSnapshot | None = None,
) -> RuntimeHealthRecord:
    return RuntimeHealthRecord(
        runtime=runtime,
        window_start=window_start,
        window_end=window_end,
        events=events or {},
        warnings=warnings or {},
        latest_metadata=metadata,
    )


def build_integration_coverage_record(
    integration: str,
    *,
    providers: Iterable[str],
    status: str,
    scopes: IntegrationScopeBreakdown,
    accounts: IntegrationAccountSummary,
    coverage_ratio: float | None,
    last_success: datetime | None,
    evaluated_at: datetime,
) -> IntegrationCoverageRecord:
    return IntegrationCoverageRecord(
        integration=integration,
        providers=list(providers),
        status=status,
        scopes=scopes,
        accounts=accounts,
        coverage_ratio=coverage_ratio,
        last_success=last_success,
        evaluated_at=evaluated_at,
    )


class StubRuntimeHealthClient(RuntimeHealthClient):
    """In-memory runtime health client returning pre-canned records."""

    _records: Sequence[RuntimeHealthRecord]

    def __init__(self, records: Sequence[RuntimeHealthRecord]) -> None:
        # type: ignore[call-arg] - parent expects AsyncSession but stubbing for tests
        self._records = list(records)

    async def summarize(self, *, hours: int = 24):  # type: ignore[override]
        return list(self._records)


class StubIntegrationCoverageClient(IntegrationCoverageClient):
    """In-memory integration coverage client returning static responses."""

    _records: Sequence[IntegrationCoverageRecord]

    def __init__(self, records: Sequence[IntegrationCoverageRecord]) -> None:
        # type: ignore[call-arg]
        self._records = list(records)

    async def summarize(
        self,
        *,
        provider_mapping=None,  # type: ignore[override]
        stale_seconds=None,
    ):
        return list(self._records)


__all__ = [
    "StubRuntimeHealthClient",
    "StubIntegrationCoverageClient",
    "build_runtime_health_record",
    "build_integration_coverage_record",
]

File: sdk/ts/src/testing/analytics.ts
import {
  IntegrationAccountSummary,
  IntegrationCoverageRecord,
  IntegrationScopeBreakdown,
  RuntimeEventAggregate,
  RuntimeHealthRecord,
  RuntimeHealthSummary,
  RuntimeMetadataSnapshot,
} from "../types.js";

export function buildRuntimeHealthRecord(
  runtime: string,
  options: {
    windowStart: Date;
    windowEnd: Date;
    events?: Record<string, RuntimeEventAggregate>;
    warnings?: Record<string, RuntimeEventAggregate>;
    metadata?: RuntimeMetadataSnapshot | null;
  },
): RuntimeHealthRecord {
  return {
    runtime,
    windowStart: options.windowStart,
    windowEnd: options.windowEnd,
    events: options.events ?? {},
    warnings: options.warnings ?? {},
    latestMetadata: options.metadata ?? null,
  };
}

export function buildIntegrationCoverageRecord(options: {
  integration: string;
  providers: string[];
  status: string;
  scopes: IntegrationScopeBreakdown;
  accounts: IntegrationAccountSummary;
  coverageRatio: number | null;
  lastSuccess: Date | null;
  evaluatedAt: Date;
}): IntegrationCoverageRecord {
  return {
    integration: options.integration,
    providers: [...options.providers],
    status: options.status,
    scopes: options.scopes,
    accounts: options.accounts,
    coverageRatio: options.coverageRatio,
    lastSuccess: options.lastSuccess,
    evaluatedAt: options.evaluatedAt,
  };
}

export class StubRuntimeHealthClient {
  constructor(
    private readonly records: RuntimeHealthRecord[],
    private readonly windowHours = 24,
    private readonly generatedAt: Date = new Date(),
  ) {}

  async getRuntimeHealth(): Promise<RuntimeHealthSummary> {
    return {
      windowHours: this.windowHours,
      generatedAt: this.generatedAt,
      runtimes: [...this.records],
    };
  }
}

export class StubIntegrationCoverageClient {
  constructor(private readonly records: IntegrationCoverageRecord[]) {}

  async getCoverage(): Promise<IntegrationCoverageRecord[]> {
    return [...this.records];
  }
}

export default {
  StubRuntimeHealthClient,
  StubIntegrationCoverageClient,
  buildRuntimeHealthRecord,
  buildIntegrationCoverageRecord,
};
